### PR TITLE
Improve encode c api

### PR DESCRIPTION
**Pull request by sz3**
The flip side of https://github.com/sz3/libcimbar/pull/144: *complex-ify* the encoder. :melting_face: 

We (users of the cimbar.js encoder, including me) have run into situations where the browser hits an out of memory (OOM) error on large input files. Example issues:
https://github.com/sz3/libcimbar/issues/141
https://github.com/sz3/libcimbar/issues/132
https://github.com/sz3/libcimbar/issues/118
https://github.com/sz3/libcimbar/issues/96

While I can't solve all OOM issues -- wirehair needs the compressed representation stored in contiguous memory -- writing the decoder (#134) reminded/convinced me that we can do better when loading the *uncompressed* data for encoding.

So:
* we're now loading input files in chunks
* this requires us to change the c api for encoding
* as a result, OOM errors in the browser should be less frequent. I won't say solved; but, better. 

The new API surface:
* `cimbare_init_encode(filename, encode_id)` -> starts an encode, waits for data
* `cimbare_encode(data)` -> provides data. When the size of data provided is == `cimbare_encode_bufsize()`, we treat this as an incremental write and don't finalize anything (and return `1`). Otherwise, we consider the write complete.
* `cimbare_encode(nullptr, 0)` therefore functions as a sort of "flush()" operation if you want to make sure the data is ready.
* when cimbare_encode() returns `0`, we can start doing our `cimbare_next_frame()` calls

### CODE REVIEW

File: src/lib/cimbar_js/cimbar_js.cpp [hunk 1]
```diff
@@ -117,17 +127,51 @@ int cimbare_encode(const unsigned char* buffer, unsigned size, const char* filen
 		return -5;
 	}
 
-	Encoder enc;
 	if (encode_id < 0)
-		enc.set_encode_id(++_encodeId); // increment _encodeId every time we change files
+		++_encodeId; // increment _encodeId every time we change files
 	else
-		enc.set_encode_id(static_cast<uint8_t>(encode_id));
+		_encodeId = encode_id;
```
Comment (sz3) [inline src/lib/cimbar_js/cimbar_js.cpp:133]: This might be the strangest side effect of this change -- `_encodeId` will now persist if you override it, so you can change the auto-increment starting point. You probably shouldn't, though. The right behavior will generally be to use `-1` (for auto-increment).

(should probably define a constant for that)

File: src/lib/cimbar_js/cimbar_js.cpp [hunk 2]
```diff
@@ -117,17 +127,51 @@ int cimbare_encode(const unsigned char* buffer, unsigned size, const char* filen
...
+
+	// otherwise, we're ready
+	unsigned fountainChunkSize = cimbar::Config::fountain_chunk_size();
+	size_t compressedSize = _comp->size();
+	if (compressedSize < fountainChunkSize)
+		_comp->pad(fountainChunkSize - compressedSize + 1);
+
+	// create the encoder stream
+	_fes = fountain_encoder_stream::create(*_comp, fountainChunkSize, _encodeId);
+	_comp.reset();
```
Comment (sz3) [inline src/lib/cimbar_js/cimbar_js.cpp:172]: After we return 0 from `cimbare_encode`, the next call to it will return `-1` (as if we never called `cimbare_init_encode`)

File: src/lib/cimbar_js/test/cimbar_jsTest.cpp [hunk 3]
```diff
@@ -65,11 +66,30 @@ TEST_CASE( "cimbar_jsTest/testRoundtrip", "[unit]" )
...
+	std::string contents = random_string(size);
+	std::string filename = "/tmp/foobar-c语言版.txt";
+	assertEquals( 0, cimbare_init_encode(filename.data(), filename.size(), 100) );
+	assertEquals( 1, cimbare_encode(reinterpret_cast<unsigned char*>(contents.data()), contents.size()) );
+	assertEquals( -1, cimbare_next_frame() );
+
+	assertEquals( 0, cimbare_encode(nullptr, 0) );
+	assertEquals( 1, cimbare_next_frame() );
+
+	assertEquals( -1, cimbare_encode(nullptr, 0) );
```
Comment (sz3) [inline src/lib/cimbar_js/test/cimbar_jsTest.cpp:93]: Demonstrating the corner case.
